src/main/java/jenkins/plugins/git/GitSCMFileSystem.java

```diff
@@ -286,40 +289,80 @@ public boolean supportsDescriptor(SCMSourceDescriptor descriptor) {
         }
 
         static class HeadNameResult {
-            final String headName;
-            final String prefix;
-
-            private HeadNameResult(String headName, String prefix) {
-                this.headName = headName;
-                this.prefix = prefix;
+            final String remoteHeadName;
+            final String refspec;
+            final SCMRevision rev;
+
+            private HeadNameResult(String remoteHeadName, String refspec, SCMRevision rev) {
+                this.remoteHeadName = remoteHeadName;
+                this.refspec = refspec;
+                this.rev = rev;
             }
 
             static HeadNameResult calculate(@NonNull BranchSpec branchSpec,
                                             @CheckForNull SCMRevision rev,
-                                            @CheckForNull EnvVars env) {
+                                            @CheckForNull String refSpec,
+                                            @CheckForNull EnvVars env,
+                                            @CheckForNull String remoteName) {
+
                 String branchSpecExpandedName = branchSpec.getName();
                 if (env != null) {
                     branchSpecExpandedName = env.expand(branchSpecExpandedName);
                 }
+                String refspecExpandedName = refSpec;
+                if (env != null && refspecExpandedName != null) {
+                    refspecExpandedName = env.expand(refspecExpandedName);
+                }
 
+                // default to a branch (refs/heads)
                 String prefix = Constants.R_HEADS;
+                // check for a tag
                 if (branchSpecExpandedName.startsWith(Constants.R_TAGS)) {
                     prefix = Constants.R_TAGS;
+                } else {
+                    // check for FETCH_HEAD
+                    if (branchSpecExpandedName.equals(Constants.FETCH_HEAD) && refspecExpandedName != null &&
+                            !refspecExpandedName.equals("")) {
```

Comment:
We already have `StringUtils` imported, might as well use it
```suggestion
                    if (branchSpecExpandedName.equals(Constants.FETCH_HEAD) && StringUtils.isNotBlank(refSpecExpandedName)) {
```